File: backend/routing.py
"""
Routing engine (routing.py) — Person B.

Turns a scored graph into two comparable routes: the calmest path and the
shortest one. Contains NO scoring logic. If you find a threshold or a weight in
this file, it belongs in scoring.py.

The graph is DIRECTED. Every block is added twice, once per direction of
travel, because walking Bourke Street eastbound and westbound cost different
amounts once counterflow is in play.

Public surface used by main.py:
    plan(origin, destination, tolerance, loads) -> dict
    node_list() -> list[dict]
"""

from __future__ import annotations

import networkx as nx

import grid
import scoring

# Average walking pace, metres per minute. Used only for the time estimate
# shown to the user; it plays no part in route selection.
WALK_M_PER_MIN = 80.0

# In the lowest-peak search, an unmeasured block is this many times more
# expensive than a measured one of the same length. It does not change which
# ceiling is achievable - only which of the tied paths is returned - so it
# steers the result toward blocks we can actually vouch for.
UNVERIFIED_TIEBREAK = 2.5


# --------------------------------------------------------------------------
# Graph construction
# --------------------------------------------------------------------------

def scored_graph(tolerance: int, loads) -> nx.DiGraph:
    """Directed copy of the street grid, weighted by sensory cost."""
    g = nx.DiGraph()
    for a, b in grid.GRAPH.edges:
        for u, v in ((a, b), (b, a)):
            g.add_edge(u, v,
                       sensory=scoring.edge_cost(u, v, tolerance, loads),
                       length=grid.edge_length(u, v))
    return g


def lowest_peak_path(origin: str, destination: str, loads) -> list[str] | None:
    """Bottleneck shortest path: minimise the WORST block, not the total.

    plan() otherwise minimises the SUM of block costs, but a route is RATED by
    its peak. Those are different objectives, and they disagree: a path can win
    on total by threading through many cheap unmeasured blocks while still
    crossing one busy corridor. For a sensory-sensitive walker the worst block
    is the one that ruins the trip, so this offers the path that minimises it.

    Binary search over the observed densities: for a candidate ceiling, keep
    only blocks at or below it and ask whether a path still exists.
    """
    assumed = scoring.assumed_density(loads)
    levels = sorted({d for d in (scoring.block_density(u, v, loads)
                                 for u, v in grid.GRAPH.edges) if d is not None}
                    | {assumed})
    if not levels:
        return None

    best = None
    lo, hi = 0, len(levels) - 1
    while lo <= hi:
        mid = (lo + hi) // 2
        ceiling = levels[mid]
        h2 = nx.Graph()
        for u, v in grid.GRAPH.edges:
            d = scoring.block_density(u, v, loads)
            # An unmeasured block counts as the assumed density, the same
            # figure the cost function uses. Treating "unknown" as "passable at
            # any ceiling" would let this route escape into pure ignorance and
            # come back 0% measured, which is not a calmer route - it is an
            # unverified one.
            unknown = d is None
            if unknown:
                d = assumed
            if d <= ceiling:
                # Among paths that tie on the ceiling, prefer VERIFIED blocks.
                # Without this the search returns whichever path is shortest,
                # which tends to be a chain of unmeasured blocks - a route we
                # cannot vouch for, reported as 0% measured.
                w = grid.edge_length(u, v) * (UNVERIFIED_TIEBREAK if unknown else 1.0)
                h2.add_edge(u, v, cost=w)
        if h2.has_node(origin) and h2.has_node(destination) \
                and nx.has_path(h2, origin, destination):
            best = nx.shortest_path(h2, origin, destination, weight="cost")
            hi = mid - 1
        else:
            lo = mid + 1
    return best


# --------------------------------------------------------------------------
# Summarising a path
# --------------------------------------------------------------------------

def _summarise(path: list[str], route_id: str, tolerance: int, loads) -> dict:
    blocks = list(zip(path, path[1:]))
    densities = [scoring.block_density(u, v, loads) for u, v in blocks]
    measured = [d for d in densities if d is not None]

    total_m = sum(grid.edge_length(u, v) for u, v in blocks)
    peak = max(measured) if measured else None

    # A route's rating is driven by its WORST block, not its average. One bad
    # corridor is what ruins a trip for a sensory-sensitive walker.
    rating = scoring.rate(peak)

    return {
        "id": route_id,
        "path": path,
        "coords": [list(c) for c in grid.coords(path)],
        "blocks": len(blocks),
        "distance_m": round(total_m),
        "minutes": round(total_m / WALK_M_PER_MIN),
        "rating": rating,
        "peak_density": peak,
        "mean_density": round(sum(measured) / len(measured)) if measured else None,
        "coverage_pct": round(100 * len(measured) / len(blocks)) if blocks else 0,
        "steps": [
            {
                "from": u, "to": v,
                "street": grid.edge_street(u, v),
                "length_m": grid.edge_length(u, v),
                "density": d,
                "rating": scoring.rate(d),
                "measured": d is not None,
            }
            for (u, v), d in zip(blocks, densities)
        ],
    }


# --------------------------------------------------------------------------
# Public API
# --------------------------------------------------------------------------

def _entry_options(spec, tolerance: int, loads) -> list[tuple[str, float, dict | None]]:
    """Ways to join the street network from a start/end spec.

    A named intersection has exactly one, costing nothing. A projected position
    has two - either end of the block the person is standing on - and the router
    picks whichever gives the better trip overall, rather than assuming the
    nearer corner is the right one.

    Returns (node, walk_metres, access_leg_or_None).
    """
    if isinstance(spec, str):
        return [(spec, 0.0, None)]

    out = []
    for end, along in (("u", "to_u_m"), ("v", "to_v_m")):
        node = spec[end]
        walk = spec["offset_m"] + spec[along]
        out.append((node, walk, {
            "to_node": node,
            "metres": round(walk, 1),
            "minutes": round(walk / WALK_M_PER_MIN),
            "offset_m": spec["offset_m"],
            "along_block_m": spec[along],
            "on_block": f"{spec['u']} -> {spec['v']}",
            "point": list(spec["point"]),
        }))
    return out


def _access_cost(metres: float, tolerance: int, loads) -> float:
    """Price the walk-in on the same scale as a block, using the assumed
    density, so it competes fairly against the routed portion."""
    if metres <= 0:
        return 0.0
    assumed = min(scoring.assumed_density(loads), scoring.DENSITY_CEILING)
    return metres * (1.0 + scoring.W_DENSITY * assumed / max(tolerance, 1))


def plan(origin, destination, tolerance: int, loads,
         source: str = "unknown") -> dict:
    """Two routes between two intersections: calmest and shortest.

    Raises ValueError for unknown intersection names, which main.py turns into
    a 400 rather than a 500.
    """
    for spec, label in ((origin, "origin"), (destination, "destination")):
        if isinstance(spec, str) and spec not in grid.NODES:
            raise ValueError(f"unknown intersection: {spec}")

    o_opts = _entry_options(origin, tolerance, loads)
    d_opts = _entry_options(destination, tolerance, loads)

    g = scored_graph(tolerance, loads)

    # Try every combination of entry and exit corner, and keep the cheapest
    # overall. This is why projecting beats snapping: the nearer corner is not
    # always the one that gives the better walk.
    best = None
    for o_node, o_walk, o_leg in o_opts:
        for d_node, d_walk, d_leg in d_opts:
            if o_node == d_node:
                continue
            try:
                p = nx.shortest_path(g, o_node, d_node, weight="sensory")
            except nx.NetworkXNoPath:
                continue
            total = (sum(g[a][b]["sensory"] for a, b in zip(p, p[1:]))
                     + _access_cost(o_walk, tolerance, loads)
                     + _access_cost(d_walk, tolerance, loads))
            if best is None or total < best[0]:
                best = (total, o_node, d_node, o_leg, d_leg, o_walk, d_walk)

    if best is None:
        raise ValueError("no route between those points")
    _, origin, destination, o_leg, d_leg, o_walk, d_walk = best

    if origin == destination:
        raise ValueError("origin and destination are the same")

    calm_path = nx.shortest_path(g, origin, destination, weight="sensory")
    fast_path = nx.shortest_path(g, origin, destination, weight="length")

    quiet_path = lowest_peak_path(origin, destination, loads)

    calm = _summarise(calm_path, "calm", tolerance, loads)
    routes = [calm]
    seen = [calm_path]

    # Lowest-peak route: what the RATING actually measures. Often different
    # from the lowest-total route, and usually the one a sensory-sensitive
    # walker would pick.
    if quiet_path and quiet_path not in seen:
        routes.append(_summarise(quiet_path, "quiet", tolerance, loads))
        seen.append(quiet_path)

    # Only offer the fast route when it is genuinely a different option.
    if fast_path not in seen:
        routes.append(_summarise(fast_path, "fast", tolerance, loads))

    # Confidence is computed, not decorative: it degrades when much of the
    # route is unmeasured or when the crowd data is a cached fallback.
    cov = calm["coverage_pct"]
    if source == "cached" or cov < 40:
        confidence = "low"
    elif cov < 70:
        confidence = "medium"
    else:
        confidence = "high"

    access_m = o_walk + d_walk
    for r in routes:
        r["access_m"] = round(access_m)
        r["total_m"] = round(r["distance_m"] + access_m)
        r["total_minutes"] = round(r["total_m"] / WALK_M_PER_MIN)

    return {
        "data": {"routes": routes,
                 "access": {"start": o_leg, "end": d_leg}},
        "meta": {
            "source": source,
            "confidence": confidence,
            "basis": f"{cov}% of the calm route has live sensor coverage",
            "coverage_pct": cov,
            "tolerance": tolerance,
            "attribution": "City of Melbourne, CC BY 4.0",
        },
    }


def node_list() -> list[dict]:
    """Every intersection, for the origin/destination pickers in the client."""
    return [{"id": n, "lat": ll[0], "lng": ll[1]} for n, ll in sorted(grid.NODES.items())]


def blocks(loads) -> dict:
    """Every block the router knows about, with its current reading.

    A debug view: it answers "does the network actually contain the street I am
    standing on, and what does it think is happening there?" without having to
    plan a route to find out.
    """
    out = []
    for u, v, d in grid.GRAPH.edges(data=True):
        density = scoring.block_density(u, v, loads)
        sids = grid.sensors_for_edge(u, v)
        out.append({
            "from": u, "to": v,
            "street": d["street"],
            "little": d["street"] in grid.LITTLE_NAMES,
            "coords": [list(grid.NODES[u]), list(grid.NODES[v])],
            "length_m": d["length_m"],
            "density": density,
            "rating": scoring.rate(density),
            "sensors": sids,
            "sensor_names": [grid.SENSORS[s].description for s in sids],
        })
    return {
        "data": {"blocks": out},
        "meta": {
            "blocks": len(out),
            "measured": sum(1 for b in out if b["density"] is not None),
            "little_streets": grid.INCLUDE_LITTLE_STREETS,
            "match_radius_m": grid.SENSOR_RADIUS_M,
        },
    }


def heatmap(loads) -> dict:
    """Current density at every reporting sensor, normalised for the client.

    Sensors that did not report are omitted entirely rather than sent as zero,
    so the client can render a gap as a gap.
    """
    points = []
    for sid, s in grid.SENSORS.items():
        load = loads.get(sid)
        if load is None:
            continue
        points.append({
            "id": sid,
            "lat": s.lat, "lng": s.lng,
            "density": load.total,
            "d1": load.d1,
            "d2": load.d2,
            "intensity": scoring.normalise(load.total),
            "rating": scoring.rate(load.total),
            "name": s.description,
        })
    return {
        "data": {"points": points},
        "meta": {
            "reporting": len(points),
            "sensors": len(grid.SENSORS),
            "attribution": "City of Melbourne, CC BY 4.0",
        },
    }

def find_refuges(lat: float | None = None, lng: float | None = None, 
                 node: str | None = None, radius_m: float = 800.0) -> dict:
    if node and node in grid.NODES:
        lat, lng = grid.NODES[node]
    elif lat is None or lng is None:
        lat, lng = -37.8136, 144.9631  # CBD default center

    refuges = grid.find_nearby_refuges(lat, lng, radius_m=radius_m)
    return {
        "data": {"refuges": refuges},
        "meta": {"count": len(refuges), "radius_m": radius_m}
    }
# --------------------------------------------------------------------------
# Self-test
# --------------------------------------------------------------------------

if __name__ == "__main__":
    import json
    import crowd

    loads = crowd.stub_loads(grid.SENSORS)
    result = plan("Spencer/Flinders", "Spring/LaTrobe", tolerance=80,
                  loads=loads, source=crowd.source())

    for r in result["data"]["routes"]:
        print(f"{r['id']:>5}: {r['blocks']} blocks, {r['distance_m']} m, "
              f"{r['minutes']} min, rating {r['rating']}, "
              f"peak {r['peak_density']}, coverage {r['coverage_pct']}%")

    print(f"\nmeta: {result['meta']['confidence']} confidence "
          f"— {result['meta']['basis']}")

    print("\nfirst four steps of the calm route:")
    for s in result["data"]["routes"][0]["steps"][:4]:
        tag = f"{s['density']}/min" if s["measured"] else "unmeasured"
        print(f"   {s['street']:<12} {s['from']:>22} -> {s['to']:<22} "
              f"{s['rating']:>7}  {tag}")

    hm = heatmap(loads)
    print(f"\nheatmap: {hm['meta']['reporting']} of {hm['meta']['sensors']} "
          f"sensors reporting")
    print(f"nodes available to the client: {len(node_list())}")

    print("\nerror handling:")
    for bad in ("Nowhere/Street", "Spencer/Flinders"):
        try:
            plan("Spencer/Flinders", bad, 80, loads)
            print(f"   {bad}: no error raised")
        except ValueError as e:
            print(f"   {bad}: ValueError — {e}")

File: frontend/src/services/engine/routing.js
/**
 * Routing engine — port of backend/routing.py.
 *
 * Turns a scored graph into comparable routes: the calmest path, the
 * lowest-peak path, and the shortest one. Contains NO scoring logic — a
 * threshold or a weight belongs in scoring.js.
 *
 * The graph is DIRECTED: every block is added twice, once per direction of
 * travel, because walking a street one way and back can cost different
 * amounts once counterflow is in play.
 */

import * as grid from './grid.js'
import * as scoring from './scoring.js'

// Average walking pace, metres per minute — used only for the time estimate.
const WALK_M_PER_MIN = 80.0

// In the lowest-peak search, an unmeasured block costs this many times more
// than a measured one of the same length, steering ties toward blocks we can
// actually vouch for.
const UNVERIFIED_TIEBREAK = 2.5

// --------------------------------------------------------------------------
// Generic Dijkstra — the grid is small (under 200 directed edges), so a
// linear scan beats the bookkeeping of a real binary heap.
// --------------------------------------------------------------------------

function dijkstra(adj, source, target, weightFn) {
  if (source === target) return [source]
  if (!adj.has(source) || !adj.has(target)) return null

  const dist = new Map([[source, 0]])
  const prev = new Map()
  const visited = new Set()
  const queue = new Set([source])

  while (queue.size) {
    let current = null
    let currentDist = Infinity
    for (const id of queue) {
      const d = dist.get(id) ?? Infinity
      if (d < currentDist) { currentDist = d; current = id }
    }
    if (current === null) break

    queue.delete(current)
    visited.add(current)
    if (current === target) break

    for (const [next, data] of adj.get(current) ?? []) {
      if (visited.has(next)) continue
      const candidate = currentDist + weightFn(data)
      if (candidate < (dist.get(next) ?? Infinity)) {
        dist.set(next, candidate)
        prev.set(next, current)
        queue.add(next)
      }
    }
  }

  if (!prev.has(target)) return null
  const path = [target]
  let cursor = target
  while (cursor !== source) {
    cursor = prev.get(cursor)
    if (cursor === undefined) return null
    path.unshift(cursor)
  }
  return path
}

function samePath(a, b) {
  return a.length === b.length && a.every((n, i) => n === b[i])
}

// --------------------------------------------------------------------------
// Graph construction
// --------------------------------------------------------------------------

/** Directed copy of the street grid, weighted by sensory cost. */
function scoredGraph(tolerance, loads) {
  const g = new Map()
  for (const name of grid.NODES.keys()) g.set(name, new Map())
  for (const [a, b] of grid.EDGES) {
    for (const [u, v] of [[a, b], [b, a]]) {
      g.get(u).set(v, {
        sensory: scoring.edgeCost(u, v, tolerance, loads),
        length: grid.edgeLength(u, v),
      })
    }
  }
  return g
}

/**
 * Bottleneck shortest path: minimise the WORST block, not the total. A route
 * is RATED by its peak, so this offers the path a sensory-sensitive walker
 * would actually want — binary search over observed densities, keeping only
 * blocks at or below the candidate ceiling and asking whether a path exists.
 */
function lowestPeakPath(origin, destination, loads) {
  const assumed = scoring.assumedDensity(loads)
  const levelSet = new Set([assumed])
  for (const [u, v] of grid.EDGES) {
    const d = scoring.blockDensity(u, v, loads)
    if (d != null) levelSet.add(d)
  }
  const levels = [...levelSet].sort((a, b) => a - b)
  if (levels.length === 0) return null

  let best = null
  let lo = 0
  let hi = levels.length - 1
  while (lo <= hi) {
    const mid = Math.floor((lo + hi) / 2)
    const ceiling = levels[mid]

    const h2 = new Map()
    const ensure = (n) => { if (!h2.has(n)) h2.set(n, new Map()) }
    for (const [u, v] of grid.EDGES) {
      let d = scoring.blockDensity(u, v, loads)
      const unknown = d == null
      if (unknown) d = assumed
      if (d <= ceiling) {
        // Among paths tying on the ceiling, prefer VERIFIED blocks — without
        // this the search returns whichever path is shortest, which tends
        // to be a chain of unmeasured blocks we cannot vouch for.
        const w = grid.edgeLength(u, v) * (unknown ? UNVERIFIED_TIEBREAK : 1.0)
        ensure(u); ensure(v)
        h2.get(u).set(v, w)
        h2.get(v).set(u, w)
      }
    }

    const path = h2.has(origin) && h2.has(destination)
      ? dijkstra(h2, origin, destination, (w) => w)
      : null
    if (path) { best = path; hi = mid - 1 } else { lo = mid + 1 }
  }
  return best
}

// --------------------------------------------------------------------------
// Summarising a path
// --------------------------------------------------------------------------

function summarise(path, routeId, tolerance, loads, predicted) {
  const blockPairs = []
  for (let i = 0; i < path.length - 1; i++) blockPairs.push([path[i], path[i + 1]])
  const densities = blockPairs.map(([u, v]) => scoring.blockDensity(u, v, loads))
  const measured = densities.filter((d) => d != null)

  const totalM = blockPairs.reduce((s, [u, v]) => s + grid.edgeLength(u, v), 0)
  // A route's rating is driven by its WORST block, not its average — one
  // bad corridor is what ruins a trip for a sensory-sensitive walker.
  const peak = measured.length ? Math.max(...measured) : null

  return {
    id: routeId,
    path,
    coords: grid.coords(path),
    blocks: blockPairs.length,
    distance_m: Math.round(totalM),
    minutes: Math.round(totalM / WALK_M_PER_MIN),
    rating: scoring.rate(peak, predicted),
    peak_density: peak,
    mean_density: measured.length ? Math.round(measured.reduce((a, b) => a + b, 0) / measured.length) : null,
    coverage_pct: blockPairs.length ? Math.round((100 * measured.length) / blockPairs.length) : 0,
    steps: blockPairs.map(([u, v], i) => ({
      from: u,
      to: v,
      street: grid.edgeStreet(u, v),
      length_m: grid.edgeLength(u, v),
      density: densities[i],
      rating: scoring.rate(densities[i], predicted),
      measured: densities[i] != null,
    })),
  }
}

// --------------------------------------------------------------------------
// Public API
// --------------------------------------------------------------------------

/**
 * Ways to join the street network from a start/end spec. A named
 * intersection has exactly one, costing nothing. A projected position has
 * two — either end of the block the person is standing on — and the router
 * picks whichever gives the better trip overall.
 */
function entryOptions(spec) {
  if (typeof spec === 'string') return [[spec, 0.0, null]]

  const out = []
  for (const [end, along] of [['u', 'to_u_m'], ['v', 'to_v_m']]) {
    const node = spec[end]
    const walk = spec.offset_m + spec[along]
    out.push([node, walk, {
      to_node: node,
      metres: Math.round(walk * 10) / 10,
      minutes: Math.round(walk / WALK_M_PER_MIN),
      offset_m: spec.offset_m,
      along_block_m: spec[along],
      on_block: `${spec.u} -> ${spec.v}`,
      point: spec.point,
    }])
  }
  return out
}

/** Price the walk-in on the same scale as a block, using the assumed density. */
function accessCost(metres, tolerance, loads) {
  if (metres <= 0) return 0.0
  const assumed = Math.min(scoring.assumedDensity(loads), scoring.DENSITY_CEILING)
  return metres * (1.0 + (scoring.W_DENSITY * assumed) / Math.max(tolerance, 1))
}

/**
 * Routes between two intersections: calmest, lowest-peak, and shortest.
 * `origin`/`destination` are either a node name (string) or a projected
 * spec from `grid.projectToEdge`. Throws for unknown intersection names or
 * when no route exists.
 */
export function plan(origin, destination, tolerance, loads, source = 'unknown') {
  const predicted = source === 'predicted'
  for (const spec of [origin, destination]) {
    if (typeof spec === 'string' && !grid.NODES.has(spec)) {
      throw new Error(`unknown intersection: ${spec}`)
    }
  }

  const oOpts = entryOptions(origin)
  const dOpts = entryOptions(destination)
  const g = scoredGraph(tolerance, loads)

  // Try every combination of entry and exit corner, keep the cheapest
  // overall — this is why projecting beats snapping: the nearer corner is
  // not always the one that gives the better walk.
  let best = null
  for (const [oNode, oWalk, oLeg] of oOpts) {
    for (const [dNode, dWalk, dLeg] of dOpts) {
      if (oNode === dNode) continue
      const p = dijkstra(g, oNode, dNode, (d) => d.sensory)
      if (!p) continue
      let total = 0
      for (let i = 0; i < p.length - 1; i++) total += g.get(p[i]).get(p[i + 1]).sensory
      total += accessCost(oWalk, tolerance, loads) + accessCost(dWalk, tolerance, loads)
      if (best === null || total < best.total) best = { total, oNode, dNode, oLeg, dLeg, oWalk, dWalk }
    }
  }

  if (best === null) throw new Error('no route between those points')
  const { oNode: originNode, dNode: destNode, oLeg, dLeg, oWalk, dWalk } = best
  if (originNode === destNode) throw new Error('origin and destination are the same')

  const calmPath = dijkstra(g, originNode, destNode, (d) => d.sensory)
  const fastPath = dijkstra(g, originNode, destNode, (d) => d.length)
  const quietPath = lowestPeakPath(originNode, destNode, loads)

  const calm = summarise(calmPath, 'calm', tolerance, loads, predicted)
  const routes = [calm]
  const seen = [calmPath]

  // Lowest-peak route: what the RATING actually measures, and often
  // different from the lowest-total route.
  if (quietPath && !seen.some((p) => samePath(p, quietPath))) {
    routes.push(summarise(quietPath, 'quiet', tolerance, loads, predicted))
    seen.push(quietPath)
  }
  // Only offer the fast route when it is genuinely a different option.
  if (!seen.some((p) => samePath(p, fastPath))) {
    routes.push(summarise(fastPath, 'fast', tolerance, loads, predicted))
  }

  // Confidence degrades when much of the route is unmeasured or the crowd
  // data is a cached fallback.
  const cov = calm.coverage_pct
  let confidence
  if (source === 'cached' || cov < 40) confidence = 'low'
  else if (cov < 70) confidence = 'medium'
  else confidence = 'high'

  const accessM = oWalk + dWalk
  for (const r of routes) {
    r.access_m = Math.round(accessM)
    r.total_m = Math.round(r.distance_m + accessM)
    r.total_minutes = Math.round(r.total_m / WALK_M_PER_MIN)
  }

  return {
    data: { routes, access: { start: oLeg, end: dLeg } },
    meta: {
      source,
      confidence,
      basis: predicted
        ? `${cov}% of the calm route has historical baseline coverage`
        : `${cov}% of the calm route has live sensor coverage`,
      coverage_pct: cov,
      // What an unmeasured block is assumed to cost — the same number the
      // pathfinder itself used to price them. A route with 0% sensor
      // coverage has a `peak_density` of null; callers should treat this as
      // the honest stand-in for "unknown, don't assume it's quiet."
      assumed_density: Math.round(scoring.assumedDensity(loads)),
      tolerance,
      attribution: 'City of Melbourne, CC BY 4.0',
    },
  }
}

/** Every intersection, for origin/destination pickers. */
export function nodeList() {
  return [...grid.NODES.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([id, [lat, lng]]) => ({ id, lat, lng }))
}

/** Every block the router knows about, with its current reading. */
export function blocks(loads) {
  const out = []
  for (const [u, v] of grid.EDGES) {
    const density = scoring.blockDensity(u, v, loads)
    const nearby = grid.sensorsForEdge(u, v)
    const street = grid.edgeStreet(u, v)
    out.push({
      from: u,
      to: v,
      street,
      little: grid.LITTLE_NAMES.has(street),
      coords: [grid.NODES.get(u), grid.NODES.get(v)],
      length_m: grid.edgeLength(u, v),
      density,
      rating: scoring.rate(density),
      // Every sensor within range, not just the one that won `density` —
      // scoring.busiestSensor() picks the worst reading among these.
      sensors: nearby.map((s) => s.sensorId),
      sensor_names: nearby.map((s) => grid.SENSORS.get(s.sensorId)?.description),
    })
  }
  return {
    data: { blocks: out },
    meta: {
      blocks: out.length,
      measured: out.filter((b) => b.density != null).length,
      little_streets: grid.INCLUDE_LITTLE_STREETS,
      match_radius_m: grid.SENSOR_RADIUS_M,
    },
  }
}

/**
 * Current density at every reporting sensor, normalised for the client.
 * `predicted` swaps in the historical-baseline-calibrated scale for both
 * `intensity` (marker size) and `rating` — otherwise every predicted sensor
 * normalises to near-zero against the live ceiling and reads "low"
 * regardless of how busy it actually is relative to other predicted hours.
 */
export function heatmap(loads, predicted = false) {
  const points = []
  for (const [sid, s] of grid.SENSORS) {
    const load = loads.get(sid)
    if (load == null) continue
    points.push({
      id: sid,
      lat: s.lat,
      lng: s.lng,
      density: load.total,
      d1: load.d1,
      d2: load.d2,
      intensity: scoring.normalise(load.total, predicted),
      rating: scoring.rate(load.total, predicted),
      name: s.description,
    })
  }
  return {
    data: { points },
    meta: {
      reporting: points.length,
      sensors: grid.SENSORS.size,
      attribution: 'City of Melbourne, CC BY 4.0',
    },
  }
}

export function findRefuges({ lat = null, lng = null, node = null, radiusM = 800.0 } = {}) {
  let rlat = lat
  let rlng = lng
  if (node && grid.NODES.has(node)) [rlat, rlng] = grid.NODES.get(node)
  else if (rlat == null || rlng == null) [rlat, rlng] = [-37.8136, 144.9631]

  const refuges = grid.findNearbyRefuges(rlat, rlng, radiusM)
  return { data: { refuges }, meta: { count: refuges.length, radius_m: radiusM } }
}
